[DO NOT MERGE] Update for forthcoming beta MicroPython

microbit-matt-hillsdon opened the pull request:
API is not yet final.

Issues:
- I think the docs PRs are missing get_touches too but will revisit that. 
(Added in https://github.com/bbcmicrobit/micropython/pull/802/files)

microbit-matt-hillsdon committed:
Add was_touched

I think the docs PRs are missing get_touches too but will revisit that.

microbit-grace committed:
Add `get_touches` method docs

microbit-matt-hillsdon committed:
WIP recording playback APIs

These are not final and we have outstanding queries.

microbit-grace committed:
Change :returns: -> :return:

microbit-grace committed:
Add examples

microbit-grace committed:
Tweak

microbit-grace committed:
Update to deep_sleep doc

To match changes in docs: Update Power Mgm to change run_every behaviour #769

microbit-grace committed:
Update run_every microbit doc

To match PR docs: Add missing days parameter to microbit.run_every. (#767)

microbit-grace committed:
Change the en/typeshed instead of welsh typeshed

microbit-grace committed:
Revert "Update to deep_sleep doc"

This reverts commit 5ff1eb7aaaec8816ac9b0451ea889a6b98013a3b.
Accidentally changed lang/ca/typeshed (Welsh one) instead of en.

microbit-grace committed:
Revert "Change the en/typeshed instead of welsh typeshed"

This reverts commit 9704523fe86bfb8ea2b0e35851f0793ce27a6127.

microbit-matt-hillsdon commented on lang/en/typeshed/stdlib/microbit/microphone.pyi:
```diff
@@ -68,3 +69,74 @@ def sound_level() -> int:
     """
     ...
+
+def record(duration: int = 3000, rate: int = 7812) -> AudioFrame:
```
Duration has a default of 3000 but is required.

microbit-matt-hillsdon commented on lang/en/typeshed/stdlib/microbit/microphone.pyi:
```diff
@@ -68,3 +69,74 @@ def sound_level() -> int:
+    """Record sound into an ``AudioFrame`` for the amount of time indicated by
+    ``duration`` at the sampling rate indicated by ``rate``.
+
+    Example: ``my_frame = microphone.record()``
```
This will need to change too.

microbit-grace committed:
ticks_cpu doc

microbit-grace commented on lang/en/typeshed/stdlib/time.pyi:
```diff
@@ -56,6 +56,16 @@ def ticks_us() -> int:
     """
     ...
 
+def ticks_cpu() -> int:
+    """
+    Similar to ticks_ms and ticks_us, but with higher resolution in CPU cycles.
+
+    Example: ``time.ticks_cpu()``
+
+    :return: The counter value in CPU cycles.
+    """
+    ...
+
```
Hi @microbit-carlos, this is the best guess we made for `ticks_cpu` because we couldn't find micropython docs on it. Is this right and are there any more specific information we can add about this? Thanks!

microbit-carlos replied:
Hi Grace,

I wasn't aware that the `ticks_cpu` function was enabled in the V2 MicroPython port, do you know when it was added? Has it been added only in the latest release?

There is some upstream documentation here: https://docs.micropython.org/en/latest/library/time.html#time.ticks_cpu

But it would be good to understand the exact implementation for micro:bit and to have that in the official docs as well. I'll have a look.

microbit-matt-hillsdon replied:
Upgraded here: https://github.com/microbit-foundation/micropython-microbit-v2/commit/fdaf840e86ed4721123b8b1fddb4830e387532b2

That diff contains the implementation that uses CYCCNT.

microbit-matt-hillsdon replied:
For the moment we always return zero in the sim for ticks_cpu but we could return the same as ticks_us or a multiple. ticks_us is already a 1000 * multiple of ticks_ms which is our real resolution in the sim.

microbit-carlos replied:
Looks like it was added with the update to MicroPython 1.22.0 in January: https://github.com/microbit-foundation/micropython-microbit-v2/pull/166
It uses low level Arm register to count cycles, so this should be correct (of course, there will be a bit of overhead by calling and returning from a Python function but that's the same for every other time.ticks_xx function):
https://github.com/microbit-foundation/micropython-microbit-v2/blob/8d9067d91bcfe316cd3aea5af1a873a37ca23d13/src/codal_port/mphalport.h#L48-L55

The thing to keep in mind is that this is register is 32 bits and we run at 64 MHz, so the register overflows every minute ish ( 2^32 / 64_000_000 = 67 seconds)

microbit-grace replied:
Thanks @microbit-carlos !

microbit-carlos replied:
Correction, it's likely 30-bits due to small int size in MicroPython.

microbit-grace committed:
Add WIP audio.sound_level stub

To be updated once there is a doc for it.

microbit-grace commented on lang/en/typeshed/stdlib/microbit/audio.pyi:
```diff
@@ -28,6 +28,14 @@ def is_playing() -> bool:
     :return: ``True`` if audio is playing, otherwise ``False``."""
     ...
 
+def sound_level() -> int:
+    """Returns the average intensity of the sound played.
+
+    Example: ``audio.sound_level()``
+
+    :return: A number between 0 and 254, being the average intensity of the sound played from the most recent chunk of data."""
+    ...
+
```
WIP - To be updated once there are micropython docs

microbit-carlos replied:
Is this blocked on something needed to be updated in https://github.com/bbcmicrobit/micropython/ ?
Also, should `254` be `255`?

microbit-carlos replied:
Ah, just saw the comment on the other PR, my answer in:
- https://github.com/bbcmicrobit/micropython/pull/791#discussion_r1549521004

microbit-matt-hillsdon replied:
We'll replace these docs with the ones you write on the docs PR (or remove them if we're not shipping **audio**.sound_level). The range came from the comment in the source code but does seem a little odd.

microbit-grace replied:
Doc has been updated here: https://github.com/bbcmicrobit/micropython/pull/791/commits/87d7f75327d84afe790fdea8a88f349fdd373fe7

microbit-grace replied:
Updated here: https://github.com/microbit-foundation/micropython-microbit-stubs/pull/97/commits/70578370697bcf25df0aa857184efcfe2f8b2c52

microbit-matt-hillsdon committed:
Add stubs for AudioFrame operations

microbit-matt-hillsdon committed:
Duration doesn't have a default in practice

I've flagged this in the docs - it's possible the implementation will be
corrected instead.

microbit-grace committed:
Update set_threshold doc to copy micropython doc

microbit-grace committed:
Suggested ammendments for set_threshold doc

microbit-grace commented on lang/en/typeshed/stdlib/microbit/microphone.pyi:
```diff
@@ -53,10 +54,19 @@ def set_threshold(event: SoundEvent, value: int) -> None:
 
-    A high threshold means the event will only trigger if the sound is very loud (>= 250 in the example).
+    The ``SoundEvent.LOUD`` event is triggered when the sound level crosses the 
+    threshold from "quiet" to "loud", and the ``SoundEvent.QUIET`` event is 
+    triggered when the sound level crosses from "loud" to "quiet".
+
+    If the ``SoundEvent.LOUD`` threshold is set lower than the 
+    ``SoundEvent.QUIET`` threshold, then the ``SoundEvent.QUIET`` threshold 
+    will decrease by one unit below the ``SoundEvent.LOUD`` threshold. If the 
+    ``SoundEvent.QUIET`` threshold is set higher than the ``SoundEvent.LOUD``
+    threshold, then the ``SoundEvent.LOUD`` threshold will increase by one unit 
+    above the ``SoundEvent.QUIET`` threshold.
 
     :param event: A sound event, such as ``SoundEvent.LOUD`` or ``SoundEvent.QUIET``.
-    :param value: The threshold level in the range 0-255.
+    :param value: The threshold level in the range 0-255. Values outside this range will be clamped.
```
This is an update based on recently [merged micropython doc PR](https://github.com/bbcmicrobit/micropython/pull/804/files). 

https://github.com/microbit-foundation/micropython-microbit-stubs/pull/97/commits/f5ca4aaf4f75db225cb50cd2c1b90e7b7487d46e - I copied `set_threshold` doc from the PR
https://github.com/microbit-foundation/micropython-microbit-stubs/pull/97/commits/936b1723bb85541d66b82e3976a4477e1c7e1322 - my attempt in updating it

I have not updated the Python Editor beta PR typeshed files yet, because in previewing the python editor, the docs look quite verbose, so I am tempted to cut it down, but am not sure how  
<img width="414" alt="Screenshot 2024-04-12 at 17 17 21" src="https://github.com/microbit-foundation/micropython-microbit-stubs/assets/145345672/ab8eaf1c-372e-4e24-8c72-35f53b842fde">

microbit-grace committed:
Update audio.sound_level doc

microbit-grace committed:
Add microphone sound_level_db doc

microbit-grace commented on lang/en/typeshed/stdlib/microbit/microphone.pyi:
```diff
@@ -53,18 +54,107 @@ def set_threshold(event: SoundEvent, value: int) -> None:
     """
     ...
+
+def sound_level_db() -> int:
+    """Get the sound pressure level in decibels.
+
+    Example: ``microphone.sound_level_db()``
+
+    :return: A representation of the sound pressure level in decibels (dB).
+    """
+    ...
```
May be subject to change following https://github.com/bbcmicrobit/micropython/pull/803 PR

microbit-grace replied:
Updated here: [d7884fa](https://github.com/microbit-foundation/micropython-microbit-stubs/pull/97/commits/d7884fa8f3185480705d2a5e4100d3ead4430a90)

microbit-matt-hillsdon committed:
Initial stubs updates

We'll have to do something about dunder docs as they're really key to
this API

microbit-matt-hillsdon committed:
Corrections

microbit-matt-hillsdon committed:
Corrections

microbit-matt-hillsdon committed:
Microphone stubs

microbit-matt-hillsdon committed:
Update play stubs

microbit-matt-hillsdon committed:
Docs for audio.play

microbit-grace committed:
Update docs to sync with changes in commit:
https://github.com/bbcmicrobit/micropython/pull/791/commits/b129d161192a640b0d034e9071e3b94af6731601
- Update recording default rate from 11k to 7812
- Update AudioTrack doc to mention out-of-range values

microbit-grace committed:
Add set_rate and get_rate to audio doc, tweaks

microbit-grace committed:
Docs for touch_calibrate() in MicroBitTouchPin

To match https://github.com/bbcmicrobit/micropython/pull/802/commits/b2a0dc994fa5965befe7835cba0f6556a058a2c1

microbit-grace committed:
Docs for SoundEvent.CLAP

As added here: https://github.com/bbcmicrobit/micropython/pull/793/commits/a37074f8e78392dbb87baf30d42c2c42c848b11f

microbit-grace committed:
Update doc for microphone.sound_level_db()

As updated here: https://github.com/bbcmicrobit/micropython/pull/803/commits/477c214ed952927cb8ca3e1a331144b461f28036

microbit-grace committed:
Doc for display.rotate()

As updated in https://github.com/bbcmicrobit/micropython/pull/808/commits/2588b997c43e56661c1eae7262513fe37d3def7c
